extension/awsmiddleware/middleware_test.go

```diff
@@ -0,0 +1,169 @@
+	s3v1 "github.com/aws/aws-sdk-go/service/s3"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+const (
+	testUserAgent = "user/agent"
+	testLatency   = time.Millisecond
+)
+
+type testHandler struct {
+	id             string
+	position       HandlerPosition
+	handleRequest  func(r *http.Request)
+	handleResponse func(r *http.Response)
+	start          time.Time
+	end            time.Time
+}
+
+var _ RequestHandler = (*testHandler)(nil)
+var _ ResponseHandler = (*testHandler)(nil)
+
+func (t *testHandler) ID() string {
+	return t.id
+}
+
+func (t *testHandler) Position() HandlerPosition {
+	return t.position
+}
+
+func (t *testHandler) HandleRequest(r *http.Request) {
+	t.start = time.Now()
+	if t.handleRequest != nil {
+		t.handleRequest(r)
+	}
+}
+
+func (t *testHandler) HandleResponse(r *http.Response) {
+	t.end = time.Now()
+	if t.handleResponse != nil {
+		t.handleResponse(r)
+	}
+}
+
+func (t *testHandler) Latency() time.Duration {
+	return t.end.Sub(t.start)
+}
+
+func TestHandlerPosition(t *testing.T) {
+	testCases := []struct {
+		position HandlerPosition
+		str      string
+	}{
+		{position: After, str: "after"},
+		{position: Before, str: "before"},
+	}
+	for _, testCase := range testCases {
+		position := testCase.position
+		got, err := position.MarshalText()
+		assert.NoError(t, err)
+		assert.EqualValues(t, testCase.str, got)
+		assert.NoError(t, position.UnmarshalText(got))
+		assert.Equal(t, position, testCase.position)
+	}
+}
+
+func TestInvalidHandlerPosition(t *testing.T) {
+	position := HandlerPosition(-1)
+	got, err := position.MarshalText()
+	assert.Error(t, err)
+	assert.ErrorIs(t, err, errUnsupportedPosition)
+	assert.Nil(t, got)
+	err = position.UnmarshalText([]byte("HandlerPosition(-1)"))
+	assert.Error(t, err)
+	assert.ErrorIs(t, err, errUnsupportedPosition)
+}
+
+func TestInvalidHandlers(t *testing.T) {
+	invalidHandler := &testHandler{id: "invalid handler", position: -1}
+	testExtension := &testMiddlewareExtension{
+		requestHandlers:  []RequestHandler{invalidHandler},
+		responseHandlers: []ResponseHandler{invalidHandler},
+	}
+	// v1
+	client := awstesting.NewClient()
+	err := ConfigureSDKv1(testExtension, &client.Handlers)
+	assert.Error(t, err)
+	assert.True(t, errors.Is(err, errInvalidHandler))
+	assert.True(t, errors.Is(err, errUnsupportedPosition))
+	// v2
+	err = ConfigureSDKv2(testExtension, &awsv2.Config{})
+	assert.Error(t, err)
+	assert.True(t, errors.Is(err, errInvalidHandler))
+	assert.True(t, errors.Is(err, errUnsupportedPosition))
+}
+
+func TestConfigureSDKv1(t *testing.T) {
+	middleware, recorder, server := setup(t)
+	client := awstesting.NewClient(&awsv1.Config{
+		Region:     awsv1.String("mock-region"),
+		DisableSSL: awsv1.Bool(true),
+		Endpoint:   awsv1.String(server.URL),
+	})
+	require.Equal(t, 3, client.Handlers.Build.Len())
+	require.Equal(t, 0, client.Handlers.Unmarshal.Len())
+	assert.NoError(t, ConfigureSDKv1(middleware, &client.Handlers))
```

Comment:
Should there be a test that adds multiple handlers with a mix of "before" and "after".

I am not sure I understand what "Before" and "After" will do with respect to existing middleware.

The test should verify the order they get called.